src/devices/Mlx90640/Mlx90640.cs

```diff
@@ -0,0 +1,957 @@
+        private ushort[] _outlierPixels = new ushort[5];
+        private byte _resolutionEE = 0;
+        private float _tgc = 0;
+        private short _vdd25 = 0;
+        private short _vPTAT25 = 0;
+
+        /// <summary>
+        /// MLX90640 Default I2C Address
+        /// </summary>
+        public const byte DefaultI2cAddress = 0x33;
+
+        /// <summary>
+        /// Creates a new instance of the MLX90640
+        /// </summary>
+        /// <param name="i2cDevice">The I2C device used for communication.</param>
+        public Mlx90640(I2cDevice i2cDevice)
+        {
+            _i2cDevice = i2cDevice ?? throw new ArgumentNullException(nameof(i2cDevice));
+            DumpEE();
+            ExtractVDDParameters();
+            ExtractPTATParameters();
+            ExtractVDDParameters();
+            ExtractPTATParameters();
+            ExtractGainParameters();
+            ExtractTgcParameters();
+            ExtractResolutionParameters();
+            ExtractKsTaParameters();
+            ExtractKsToParameters();
+            ExtractCPParameters();
+            ExtractAlphaParameters();
+            ExtractOffsetParameters();
+            ExtractKtaPixelParameters();
+            ExtractKvPixelParameters();
+            ExtractCILCParameters();
+            ExtractDeviatingPixels();
+        }
+
+        /// <summary>
+        /// Sets sampling to the given value
+        /// </summary>
+        /// <param name="mode">Sampling Mode</param>
+        public void SetSampling(Sampling mode)
+        {
+            Span<ushort> controlRegister1 = stackalloc ushort[1];
+
+            int value = ((byte)mode & 0x07) << 7;
+
+            controlRegister1 = ReadRegisters16(0x800D, 1);
+            value = (controlRegister1[0] & 0xFC7F) | value;
+            Write16(0x800D, value);
+        }
+
+        /// <summary>
+        /// Request both 'halves' of a frame from the sensor, merge them
+        /// and calculate the temperature in C for each of 32x24 pixels.
+        /// </summary>
+        /// <returns>Temperature of the specified pixel</returns>
+        public Temperature[] GetFrame()
+        {
+            Temperature[] framebuf = new Temperature[PixelCount];
+            float emissivity = 0.95f;
+            ushort[] mlx90640Frame = new ushort[834];
+
+            for (int i = 0; i < 2; i++)
+            {
+                if (GetFrameData(ref mlx90640Frame) < 0)
+                {
+                    throw new Exception("Frame data error");
+                }
+
+                Temperature tr = Temperature.FromDegreesCelsius(GetTa(mlx90640Frame).DegreesCelsius - OPENAIR_TA_SHIFT);
+                CalculateTo(mlx90640Frame, emissivity, tr, ref framebuf);
+            }
+
+            return framebuf;
+        }
+
+        private void CalculateTo(ushort[] frameData, float emissivity, Temperature tr, ref Temperature[] result)
+        {
+            ushort subPage = frameData[833];
```

Comment:
is there a page somewhere on the documentation that explains how to make this math?

Reply:
check **11.2.2.9. Calculating To for basic temperature range**

https://www.melexis.com/-/media/files/documents/datasheets/mlx90640-datasheet-melexis.pdf